src/script/mod.rs

```diff
@@ -115,6 +116,125 @@ impl From<opcode::Error> for Error {
     }
 }
 
+/// Evaluation functions for script components.
+pub trait Evaluable {
+    /// Get the byte length of this script sig.
+    fn byte_len(&self) -> usize;
+
+    /// Convert a sequence of `Opcode`s to the bytes that would be included in a transaction.
+    fn to_bytes(&self) -> Vec<u8>;
+
+    /// Evaluate this script component.
+    fn eval(
+        &self,
+        flags: interpreter::Flags,
+        checker: &dyn interpreter::SignatureChecker,
+        stack: interpreter::Stack<Vec<u8>>,
+    ) -> Result<interpreter::Stack<Vec<u8>>, Error>;
+
+    /// Returns true iff this script is P2SH.
+    fn is_pay_to_script_hash(&self) -> bool;
+
+    /// Called by `IsStandardTx` and P2SH/BIP62 VerifyScript (which makes it consensus-critical).
+    fn is_push_only(&self) -> bool;
+}
+
+/// A script component is used as either the script sig or script pubkey in a script. Depending on
+/// `T` (which generally needs an `opcode::Evaluable` instance), it has different properties:
+///
+/// - `PossiblyBad` – used for scripts that need to go through consensus-compatible verification
+///   (that is, read from the chain)
```

Comment:
Now that I see this documentation, it suggests that `PossiblyBad` being unnameable in the public API is the bug.